src/DataProtection/DataProtection/src/Managed/ManagedAuthenticatedEncryptor.cs

```diff
@@ -233,27 +237,33 @@ public byte[] Decrypt(ArraySegment<byte> protectedPayload, ArraySegment<byte> ad
                     }
 
                     // Step 4: Validate the MAC provided as part of the payload.
-
                     if (!CryptoUtil.TimeConstantBuffersAreEqual(correctHash, 0, correctHash.Length, protectedPayload.Array!, macOffset, eofOffset - macOffset))
                     {
                         throw Error.CryptCommon_PayloadInvalid(); // integrity check failure
                     }
 
                     // Step 5: Decipher the ciphertext and return it to the caller.
+#if NET10_0_OR_GREATER
+                    using var symmetricAlgorithm = CreateSymmetricAlgorithm(key: decryptionSubkey);
 
-                    using (var symmetricAlgorithm = CreateSymmetricAlgorithm())
-                    using (var cryptoTransform = symmetricAlgorithm.CreateDecryptor(decryptionSubkey, iv))
-                    {
-                        var outputStream = new MemoryStream();
-                        using (var cryptoStream = new CryptoStream(outputStream, cryptoTransform, CryptoStreamMode.Write))
-                        {
-                            cryptoStream.Write(protectedPayload.Array!, ciphertextOffset, macOffset - ciphertextOffset);
-                            cryptoStream.FlushFinalBlock();
+                    // note: here protectedPayload.Array is taken without an offset (cant use AsSpan() on ArraySegment)
+                    var ciphertext = protectedPayload.Array.AsSpan().Slice(ciphertextOffset, macOffset - ciphertextOffset);
+                    var iv = protectedPayload.Array.AsSpan().Slice(ivOffset, _symmetricAlgorithmBlockSizeInBytes);
 
-                            // At this point, outputStream := { plaintext }, and we're done!
-                            return outputStream.ToArray();
-                        }
+                    return symmetricAlgorithm.DecryptCbc(ciphertext, iv);
```

Comment:
are we using CBC mode? I genuinely don't know - mostly after validation / explanation of why this choice when migrating - if this is what `CryptoStream` would use, maybe link?

Also, note that since this is a private method, we might also wonder whether we can remove the decrypted buffer by using the overload that takes a `Span<byte>` input, and moving the allocation (maybe as a lease) to the consuming code site; if the input is an oversized array, then note that `MemoryStream` allows an array to be passed *in* via the .ctor, which might mean it works fine in the fallback mode too, although that is usually intended for read buffers - just make sure you use the `writable: true` param and check the length written.

I can't tell you how the size of cipher-bytes and plain-bytes relate; is it 1:1? :shrug: don't know!

but, presumably, removing this result array would have a significant impact on the remaining allocs

Reply:
thanks for suggestion: I used a new API using Spans and removing yet another allocation!